packages/uploadthing/src/effect-http.ts

```diff
@@ -0,0 +1,114 @@
+import * as Http from "@effect/platform/HttpServer";
+import * as Effect from "effect/Effect";
+
+import type { Json } from "@uploadthing/shared";
+import {
+  FetchContext,
+  getStatusCodeFromError,
+  UploadThingError,
+} from "@uploadthing/shared";
+
+import { UPLOADTHING_VERSION } from "./internal/constants";
+import { formatError } from "./internal/error-formatter";
+import {
+  buildPermissionsInfoHandler,
+  buildRequestHandler,
+} from "./internal/handler";
+import { incompatibleNodeGuard } from "./internal/incompat-node-guard";
+import { toWebRequest } from "./internal/to-web-request";
+import type { FileRouter, RouteHandlerOptions } from "./internal/types";
+import type { CreateBuilderOptions } from "./internal/upload-builder";
+import { createBuilder } from "./internal/upload-builder";
+
+export { UTFiles } from "./internal/types";
+export type { FileRouter };
+
+type MiddlewareArgs = {
+  req: Http.request.ServerRequest;
+  res: undefined;
+  event: undefined;
+};
+
+export const createUploadthing = <TErrorShape extends Json>(
+  opts?: CreateBuilderOptions<TErrorShape>,
+) => createBuilder<MiddlewareArgs, TErrorShape>(opts);
+
+export const createRouteHandler = <TRouter extends FileRouter>(
+  opts: RouteHandlerOptions<TRouter>,
+): Http.router.Router<Http.body.BodyError, never> => {
+  incompatibleNodeGuard();
+
+  const requestHandler = buildRequestHandler<TRouter, MiddlewareArgs>(
+    opts,
+    "effect-http",
+  );
+  const getBuildPerms = buildPermissionsInfoHandler<TRouter>(opts);
+
+  return Http.router.empty.pipe(
+    Http.router.get(
+      "/",
+      Http.response.json(getBuildPerms(), {
+        headers: Http.headers.fromInput({
+          "x-uploadthing-version": UPLOADTHING_VERSION,
+        }),
```

Comment:
can we provide this on a router level? All responses we return from this router should have this header. we're currently setting it 3 times

Reply:
Yes
```ts
const addUploadThingHeader = Http.middleware.make(
  Effect.map(Http.response.setHeader("x-uploadthing-version", UPLOADTHING_VERSION)),
);

Http.router.empty.pipe(
  // ...
  Http.router.use(addUploadThingHeader)
)
```
Technically `middleware.make` is just an `identity` type helper.
a `Middleware` is `<E,R>(app: App.Default<E,R>) => App.Default<any, any>`, and `App.Default` is just an alias for `Effect<ServerResponse, E,R>`. Meaning the above is just shorthand.

The `.use` would affect everything within the given `Router`. In a pipe remember that the composition is technically inverse so everything above the `.use` would be affected but not anything below it.
```ts
Http.router.empty.pipe(
  Http.router.get("a", A),
  Http.router.get("b", B),
  Http.router.get("c", C),
  Http.router.use(addUploadThingHeader)
)

// is literally
Http.router.use(addUploadThingHeader)(
  Http.router.get("c", C)(
    Http.router.get("b", B)(
      Http.router.get("a", A)(
        Http.router.empty
      )
    )
  )
)
```